transports/noise/src/lib.rs

```diff
@@ -219,6 +223,59 @@ where
     }
 }
 
+impl<T> InboundSecurityUpgrade<T> for Config
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    type Output = Output<T>;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<(PeerId, Self::Output), Self::Error>>;
+
+    fn secure_inbound(self, socket: T, _: Self::Info, _: Option<PeerId>) -> Self::Future {
+        async move {
+            let mut state = self.into_responder(socket)?;
+
+            handshake::recv_empty(&mut state).await?;
+            handshake::send_identity(&mut state).await?;
+            handshake::recv_identity(&mut state).await?;
+
+            let (pk, io) = state.finish()?;
+
+            let expected = pk.to_peer_id();
+            Ok((expected, io))
+        }
+        .boxed()
+    }
+}
+
+impl<T> OutboundSecurityUpgrade<T> for Config
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    type Output = Output<T>;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<(PeerId, Self::Output), Self::Error>>;
+
+    fn secure_outbound(self, socket: T, _: Self::Info, peer_id: Option<PeerId>) -> Self::Future {
+        async move {
+            let mut state = self.into_initiator(socket)?;
+
+            handshake::send_empty(&mut state).await?;
+            handshake::recv_identity(&mut state).await?;
+            handshake::send_identity(&mut state).await?;
+
+            let (pk, io) = state.finish()?;
+
+            let expected = pk.to_peer_id();
```

Comment:
This is a bit odd, I would have thought that the provided `PeerId` is to be named `expected` and the one we negotiated is the one we "found".

Reply:
I definitely looked at it from another perspective :-)

Reply:
When we dial an address, say `/ip4/10.0.0.1/tcp/9999/p2p/XXX` then we _expect_ the peer at the IP+port endpoint to authenticate as `XXX`.

`XXX` is what we extract from the address and pass into the `secure_outbound` function. The public key (thus `PeerId`) we receive as part of the noise handshake (`pk` above) is what the peer _actually_ authenticated with. In other words, they have control over the corresponding private key.

We need to fail the handshake if our expectation is not met. This could either be because somebody is performing a MITM attack or the address is simply wrong (e.g. bad user input). In any case, the `PeerId` coming from the _address_ is the expected `PeerId` and the one from the handshake is the actual `PeerId`.

Reply:
Thank you for your comments and excuse me for the delay in the response. Definitely the previous wording for the actual and *expected* `PeerId` was a bit confusing, so I preferred to change it to just `peer_id` for the actual one and `remote_peer_id` for the one coming from the address. I noticed that in other parts of the project the latter is already used for a similar purpose.

Reply:
> I noticed that in other parts of the project the latter is already used for a similar purpose.

Whilst we definitely use `remote_peer_id` in other parts of the project, it is not the same "purpose" as it is here. In fact, as part of a security handshake, both the _expected_ peer ID and the one the peer authenticated with as part of the handshake are **both** a "remote peer id". So it does not make any sense to use "remote" as a qualifier here.

Please change the variable names to what I suggested:

- The one provided via the parameter is the _expected_ `PeerId` (if you wanted to be super precise, you could call it `maybe_expected_peer_id` to also cover the fact that it is an `Option`)
- The one that is output as part of the auth handshake is the _actual_ `PeerId`